src/main/java/com/emc/mongoose/storage/driver/coop/netty/http/s3/S3StorageDriver.java

```diff
@@ -736,31 +811,52 @@ protected final void applyAuthHeaders(
 		if (uid == null || secret == null) {
 			return;
 		}
-		final var mac = MAC_BY_SECRET.get().computeIfAbsent(secret, GET_MAC_BY_SECRET);
-		final var canonicalForm = getCanonical(httpHeaders, httpMethod, dstUriPath);
-		final var sigData = mac.doFinal(canonicalForm.getBytes());
-		httpHeaders.set(
+
+		try {
+			if (authVersion == 2) {
+				final var mac = MAC_BY_SECRET.get().computeIfAbsent(secret, GET_MAC_BY_SECRET);
+				final var canonicalForm = getCanonical(httpHeaders, httpMethod, dstUriPath);
+				final var sigData = mac.doFinal(canonicalForm.getBytes());
+				httpHeaders.set(
 						HttpHeaderNames.AUTHORIZATION,
 						S3Api.AUTH_PREFIX + uid + ':' + BASE64_ENCODER.encodeToString(sigData));
-	}
-
-	protected String getCanonical(
-					final HttpHeaders httpHeaders, final HttpMethod httpMethod, final String dstUriPath) {
-		final var buffCanonical = BUFF_CANONICAL.get();
-		buffCanonical.setLength(0); // reset/clear
-		buffCanonical.append(httpMethod.name());
-		for (final var headerName : S3Api.HEADERS_CANONICAL) {
-			if (httpHeaders.contains(headerName)) {
-				for (final var headerValue : httpHeaders.getAll(headerName)) {
-					buffCanonical.append('\n').append(headerValue);
-				}
-			} else if (sharedHeaders != null && sharedHeaders.contains(headerName)) {
-				buffCanonical.append('\n').append(sharedHeaders.get(headerName));
 			} else {
```

Comment:
Even though we're currently supporting only 2 and 4, I'd add implicit check for the version here.
```
if (authVersion == 4) { ... }
```

Reply:
You are optimistic in terms of new auth policies release dates, but it's a fair point